[1LP][RFR] Chargeback tests for long term rates

nachandr opened the pull request:
PRT results:

58 - passed
59 - passed

dajoRH commented:
Lint report for commit 1b5cd1d583537c26f932af21a8c385e87423f52e:

`cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py`:
- :bangbang: Line 1:1: **E902** *IndentationError: unindent does not match any outer indentation level*
- :bangbang: Line 293:54: **E999** *SyntaxError: invalid syntax*
- :bangbang: Line 294:13: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 295:21: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 295:23: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 296:21: **E113** *unexpected indentation*
- :bangbang: Line 297:13: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 300:1: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 300:22: **E225** *missing whitespace around operator*
- :bangbang: Line 301:1: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 303:16: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 303:18: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 305:28: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 305:30: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 306:30: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 306:32: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 307:23: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 307:25: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 308:20: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 308:22: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 309:25: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 309:27: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 310:19: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 310:21: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 312:18: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 312:20: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 315:21: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 315:23: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 318:22: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 318:24: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 321:19: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 321:21: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 324:22: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 324:24: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 328:13: **E113** *unexpected indentation*
- :bangbang: Line 353:5: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 354:5: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 356:5: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 357:5: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 359:5: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 360:5: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 369:21: **E113** *unexpected indentation*
- :bangbang: Line 369:27: **E225** *missing whitespace around operator*
- :bangbang: Line 375:9: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 376:9: **E122** *continuation line missing indentation or outdented*
- :bangbang: Line 376:16: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 376:18: **E251** *unexpected spaces around keyword / parameter equals*
- :bangbang: Line 376:48: **E225** *missing whitespace around operator*
- :bangbang: Line 377:27: **E225** *missing whitespace around operator*

![Hmmm](https://i.imgur.com/q3iCWqE.gif)
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

RonnyPfannschmidt commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -290,10 +284,10 @@ def add_rate(tiered_rate):
                 cost = ((d['variable_rate'] * usage) + (d['fixed_rate'] * consumed_hours)) / 24
```
**OPTIONAL** as far as i an tell the formula is always the same, can we look up the divisor in a mapping

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+    try:
+        user = appliance.collections.users.create(
+            name=provider.name + fauxfactory.gen_alphanumeric(),
+            credential=Credential(principal='uid' + '{}'.format(fauxfactory.gen_alphanumeric()),
```
**REQUIRED**
Please use `format`

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+        logger.info('Assigned VM OWNERSHIP for {} running on {}'.format(vm_name, provider.name))
+
+        yield user.name
+    finally:
```
**QUESTION**
What's the point in this try-finally block?

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+    candu.enable_all()
+
+    yield
+
```
**REQUIRED**
Here and further please remove these blank lines around `yield`.

mshriver replied:
@quarckster I would disagree with this requirement, as it isn't in our style guide.  Further, I'd disagree with trying to enforce this, as I believe it falls within PEP8's guidelines for use of blank lines.  Personally I like this style for yield fixtures, or yielding methods in general, as it cleanly separates the logical blocks of the function, and makes it painfully clear that it yelds, not returns.

https://www.python.org/dev/peps/pep-0008/#blank-lines

jamesooden replied:
I too agree with @mshriver.   I was taught early on to group code kind of like we group paragraphs in human languages, using the blank lines to separate the big things we are doing.

quarckster replied:
I won't insist, but I don't like it :-P

izapolsk replied:
let me add two words from my side :). I'd ask to remove first empty line between last statement of first code block and yield since this is one logical block. however, empty line after yield is really necessary because it splits two blocks of logic and makes code more readable.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+@pytest.yield_fixture(scope="module")
+def vm_ownership(enable_candu, provider, appliance):
+    # In these tests, chargeback reports are filtered on VM owner.So,VMs have to be
+    # assigned ownership.
```
**REQUIRED**
Function doctrings use `"""` https://www.python.org/dev/peps/pep-0257

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+           record.derived_memory_used or
+           record.net_usage_rate_average or
+           record.disk_usage_rate_average):
+            return True
```
**OPTIONAL**
I think such construction is slightly better:
```python
return any((
    record.cpu_usagemhz_rate_average,
    record.cpu_usage_rate_average,
    record.derived_memory_used,
    record.net_usage_rate_average,
    record.disk_usage_rate_average
))
```

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+
+
+@pytest.fixture(scope="module")
+def resource_usage(vm_ownership, appliance, provider):
```
**REQUIRED**
This function is huge. Consider to divide it into smaller parts.

nachandr replied:
I've moved verify_records_metrics_table() outside of resource_usage(). That's the best I can do.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+        return False
+
+    wait_for(verify_records_metrics_table, [appliance, provider], timeout=600,
+        fail_condition=False, message='Waiting for VM real-time data')
```
**REQUIRED**
Here and further `fail_condition=False` is redundant.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+    report.queue(wait_for_finish=True)
+
+    yield list(report.get_saved_reports()[0].data.rows)
+    report.delete()
```
**REQUIRED**
Always check existence of an item in teardown:
```python
if report.exists:
    report.delete()
```

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+                            {'per_time': interval, 'variable_rate': '720'}})
+        storage.create()
+        yield desc
+    finally:
```
**QUESTION**
Again, I don't get the point of try/finally here.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+            assert estimated_cpu_usage_cost - DEVIATION <= float(cost) \
+                <= estimated_cpu_usage_cost + DEVIATION, \
+                'Estimated cost and report cost do not match'
+            break
```
**REQUIRED**
Here and further please use `soft_assert` fixture instead `assert` and `break`.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,473 @@
+            assert estimated_storage_usage_cost - DEVIATION <= float(cost) \
+                <= estimated_storage_usage_cost + DEVIATION, \
+                'Estimated cost and report cost do not match'
+            break
```
These tests look very similar, consider to use parametrization.

nachandr commented:
@RonnyPfannschmidt and @quarckster , I've addressed all your comments.

quarckster commented on cfme/intelligence/reports/reports.py:
```diff
@@ -310,6 +310,18 @@ def _get_state():
+        except CandidateNotFound:
+            return False
+
+    def delete_if_exists(self):
```
**REQUIRED**
This method is redundant.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+
+import fauxfactory
+import pytest
+import re
```
**REQUIRED**
```python
import fauxfactory
import math
import pytest
import re
from datetime import date
```

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+#
+# All infra and cloud providers support chargeback reports.
+# But, in order to validate costs for different rates, running the tests on just one provider
+# should suffice.
```
**REQUIRED**
Please use docstrings.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+@pytest.yield_fixture(scope="module")
+def vm_ownership(enable_candu, provider, appliance):
+    """In these tests, chargeback reports are filtered on VM owner.So,VMs have to be
+    assigned ownership."""
```
**REQUIRED**
Here and further please use multiline docstrings where it's required https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+    vm = VM.factory(vm_name, provider)
+    user = None
+    user = appliance.collections.users.create(
+        name=provider.name + '{}'.format(fauxfactory.gen_alphanumeric()),
```
**REQUIRED**
Here and further please use `format`:
```python
name='{}_{}'.format(provider.name, fauxfactory.gen_alphanumeric()),
```

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+    # are capturing C&U data and forcing hourly rollups by running these commands through
+    # the Rails console.
+
+    def verify_records_metrics_table(appliance, provider):
```
**REQUIRED**
Why did you put this function inside another function? Please take it out and place it near `verify_records_rollups_table`.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+            "\"vm = Vm.where(:ems_id => {}).where(:name => {})[0];\
+            vm.perf_capture('realtime', 1.hour.ago.utc, Time.now.utc)\""
+            .format(provider.id, repr(vm_name)))
+        assert ret.rc == 0, "Failed to capture VM C&U data:".format(ret.output)
```
**OPTIONAL**
```python
assert ret.success, "Failed to capture VM C&U data:".format(ret.output)
```

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+    cb_group = group_collection.instantiate(description='EvmGroup-user')
+
+    vm = VM.factory(vm_name, provider)
+    user = None
```
**REQUIRED**
This statement is pointless.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+            filter(details.description == metric_description).
+            filter(cb_rates.rate_type == rate_type).
+            filter(cb_rates.description == description).all()
+        )
```
**REQUIRED**
```python
result = (
    appliance.db.client.session.query(tiers)
    .join(details, tiers.chargeback_rate_detail_id == details.id)
    .join(cb_rates, details.chargeback_rate_id == cb_rates.id)
    .filter(details.description == metric_description)
    .filter(cb_rates.rate_type == rate_type)
    .filter(cb_rates.description == description).all()
)
```

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+@pytest.yield_fixture(scope="module")
+def new_compute_rate(interval):
+    """Create a new Compute Chargeback rate"""
+    desc = 'custom_' + interval
```
**REQUIRED**
Please use `format`

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,442 @@
+
+# Tests to validate costs reported in the Chargeback report for various metrics.
+# The costs reported in the Chargeback report should be approximately equal to the
+# costs estimated in the chargeback_costs_custom fixtures.
```
**REQUIRED**
It should go to doctrings.

quarckster commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,462 @@
+@pytest.yield_fixture(scope="module")
+def new_compute_rate(interval):
+    """Create a new Compute Chargeback rate"""
+    desc = 'custom_' + '{}'.format(interval)
```
**REQUIRED**
```python
desc = 'custom_{}'.format(interval)
```

nachandr commented:
@psav , I've addressed all your comments , including the one regarding skipping the tests when a report is empty and when relevant columns are missing in the report.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+    pytest.mark.provider([VMwareProvider], selector=ONE,
+                       scope='module',
+                       required_fields=[(['cap_and_util', 'test_chargeback'], True)]),
+    pytest.mark.usefixtures('has_no_providers_modscope', 'setup_provider_modscope'),
```
REQUIRED: there was email from @psav regarding provider modscope fixtures. please sort out this.

nachandr replied:
Done

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+    vm = VM.factory(vm_name, provider)
+    user = appliance.collections.users.create(
+        name='{}_{}'.format(provider.name, fauxfactory.gen_alphanumeric()),
+        credential=Credential(principal='uid' + '{}'.format(fauxfactory.gen_alphanumeric()),
```
REQUIRED: please turn this stuff into one string.

nachandr replied:
Done

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+        selections={
+            'Enterprise': {'Rate': description}
+        })
+    enterprise.computeassign()
```
OPTIONAL: it would be more readable with underscore between words.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+    ems = appliance.db.client['ext_management_systems']
+    rollups = appliance.db.client['metric_rollups']
+
+    with appliance.db.client.transaction:
```
REQUIRED: please don't use db where it is possible. we should get rid of db or ssh calls wherever it is possible. please use rest api instead.

app.rest_api.collections.metric_rollups
Out[4]: <Collection u'metric_rollups' (u'Metric Rollups')>

nachandr replied:
I'm doing inner join queries on two different tables in two places and an inner join query on three different tables in another place, hence using db.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+    metrics = appliance.db.client['metrics']
+
+    # Capture real-time C&U data
+    ret = appliance.ssh_client.run_rails_command(
```
QUESTION: are you sure there is no other way to check this w/o direct rails call ?
btw, if you really need this functionality, it makes sense to ask devs to make it available thru api.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+
+    with appliance.db.client.transaction:
+        result = (
+            appliance.db.client.session.query(metrics.id)
```
^^

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+            ems.name == provider.name, metrics.timestamp >= date.today())
+        )
+
+    for record in appliance.db.client.session.query(metrics).filter(
```
QUESTION: I see several very similar code blocks. can those be combined ?

mshriver replied:
@izapolsk this is something I had discussed with @nachandr earlier, although these look like the blocks could be replaced, the way the logic is operating the queries have to be made separately.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+    consumed_hours = 0
+    vm_name = provider.data['cap_and_util']['chargeback_vm']
+
+    metrics = appliance.db.client['metrics']
```
please ask developers to create such collection in api since it is necessary for testing

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+    appliance.server.settings.disable_server_roles(
+        'ems_metrics_coordinator', 'ems_metrics_collector')
+
+    # Perfrom rollup of C&U data.
```
typo

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+        'ems_metrics_coordinator', 'ems_metrics_collector')
+
+    # Perfrom rollup of C&U data.
+    ret = appliance.ssh_client.run_rails_command(
```
OPTIONAL: this call looks very similar to previous one. Can you combine them somehow ?
btw, if you could get rid of direct rails call, it would be even better.

nachandr replied:
It cannot be combined.

I'll reach out to Dev asking them to provide this functionality through REST API.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+            average_disk_io = average_disk_io + record.disk_usage_rate_average
+
+    # Convert storage used in Bytes to GB
+    average_storage_used = average_storage_used * math.pow(2, -30)
```
OPTIONAL: I would rather use  average_storage_used/(1024*1024*1024) because it would be readable and it wouldn't require math.

nachandr replied:
By default,chargeback rates for storage are defined in this form: 0.01 USD/GB
So, I'm converting storage used in Bytes to GB. I've added a comment regarding this.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+    """Query the DB for Chargeback rates"""
+    tiers = appliance.db.client['chargeback_tiers']
+    details = appliance.db.client['chargeback_rate_details']
+    cb_rates = appliance.db.client['chargeback_rates']
```
OPTIONAL: as far as I see some of chargeback functionality is available in api. So, it makes sense to check what is missing and ask devs to provide such functionality thru api.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+        else:
+            estimated_cpu_usage_cost = chargeback_costs_custom['cpu_used_cost']
+            cost_from_report = groups["CPU Used Cost"]
+            cost = re.sub(r'[$,]', r'', cost_from_report)
```
REQUIRED: please use string's .replace instead

quarckster replied:
Yes, with `.replace()` you don't need to import `re`.

izapolsk commented on cfme/tests/intelligence/chargeback/test_chargeback_long_term_rates.py:
```diff
@@ -0,0 +1,475 @@
+        else:
+            estimated_memory_usage_cost = chargeback_costs_custom['memory_used_cost']
+            cost_from_report = groups["Memory Used Cost"]
+            cost = re.sub(r'[$,]', r'', cost_from_report)
```
^^

mshriver commented:
marking WIPTEST due to ongoing PRT failures in 23093
